### PR TITLE
Al facilitate nested subfolder templates

### PR DESCRIPTION
**Pull request by aodl**
Related to #2640 

### Problem
Templates in the views folder can quickly stack up and become difficult to manage without the ability to group them into nested subdirectories.

#### 1st Commit
Don't hardcode `InDepth` to false in `ShapeTemplateBindingStrategy.cs`. Instead make`IShapeTemplateHarvester` expressive enough to say whether or not it points to view directories that contain nested subdirectories.

#### 2nd Commit
Not essential, but provide some overridable extension points on `BasicShapeTemplateHarvester`, to make it far easier to define new harvesters (avoiding having to rewrite error-prone code, such as that found in the `Adjust` method).

Together these changes make it simple for developers to extend the existing shape template harvester functionality, like this:

```
    public class NestedShapeTemplateHarvester : BasicShapeTemplateHarvester
    {
        public override IEnumerable<string> SubPaths() => new[] { "Views/Nested" };
        public override bool InDepth => true;
        protected override string leader(string _) => "";
    }
```
^ Haven't include the above in the PR, but could do..?

### CODE REVIEW

**Comment by sebastienros** [comment]
We were talking about your work on OC this week during triage, and I was mentioning that you reminded me a lot of @rtpHarry who was also super active with many ideas on how to change OC to improve it. Or is that just part of having the same nationality?

**Comment by aodl** [comment]
I'll take that as a compliment! :D 

I'd like to get more involved in these meetings, as you mentioned before, so that I'm up-to-date with what's going on. What time do you guys do it on Thursday? and where does it take place?

**Comment by sebastienros** [comment]
Open this url https://orchardproject.net/meeting (skype for business) on Tuesdays and Thrusdays at 12 PT (8pm london).

**Comment by sebastienros** [comment]
Could we have the `Basic` implementation be configurable using options instead? It would have a collection of paths, with default ones, and from the Startup code we could add new ones. Or provide IConfigure services to change these.

**Comment by aodl** [comment]
I thought about doing something like this, but then I realised that `IEnumerable<IShapeTemplateHarvester>` is injected into `ShapeTemplateBindingStrategy`, which is enumerated and all harvesters are used. So rather than adapting the configuration of a single instance of `IShapeTemplateHarvester`, the same functionality can be achieved by registering multiple `IShapeTemplateHarvester`s with the service provider

**Comment by sebastienros** [comment]
This is actually a very acceptable pattern to have some implementations of an interface configurable. I understand why you leaned to the solution you are providing, and you are actually following what was already done, but I still think we could use your current interest in it to make it better. This has been ported directly from O1 and we can do better in OC without much more work.

As an example, we are allowing the `X-Generator` header to be disabled with an option, instead of having to remove the service or middleware that does that.

**Comment by aodl** [comment]
Cool. I hesitated, because the way I see it this would add a new way of configuring shape template harvesters, making the solution slightly more complex, without adding any actual new functionality (in the sense that this can be done already via the existing approach, and the existing approach is more flexible because you can set `InDepth` to true for some directories and false for other etc.).

Happy to do this though as it would allow subpaths to be configured without having to write and register a new class.  

I searched the source code on the dev branch for 'X-Generator', but didn't find anything. Could you direct me to some example code where this sort of thing is being done already to give me a head start?

**Comment by aodl** [comment]
Looks good, thanks for the link. I'll push this shortly

**Comment by aodl** [comment]
Sorry for the delay on this, I became occupied with other things. I've still been thinking about this, and reading up on .Net Core configuration approaches. One thing I was wondering is if it would be considered useful to go as far as facilitating template harvester configuration to come from a config file?

### tenants.json
I thought about binding to a new configuration field/section in the `tenants.json` file, allowing different tenants to have different view folders specified. This could be useful, as it would allow a single theme to serve up different content based on the tenant (then again this may also be bad for the exact same reason, as we may want to encourage this to be done by registering different themes with each tenant - particularly as themes can inherit from another theme, so multiple themes can already share content).

### theme manifest
I also thought about allowing the configuration to be specified in the theme manifest. This seems to make a lot more sense, as its obviously a theme specific thing. 

I think I've convinced myself that the theme manifest would be the best way to go. But I wanted to put this idea up here first in case it is considered overkill, or if there is some other place for configuration that I'm not aware about that might make more sense.

**Comment by agriffard** [comment]
Conflict to solve in BasicShapeTemplateHarvester.cs

**Comment by agriffard** [comment]
@aodl Please fix the conflicts.

**Comment by aodl** [comment]
sorry for the delay on this. I've been busy with other things. I'll try and make time to resolve these conflicts soon

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapeTemplateStrategy/BasicShapeTemplateHarvester.cs [hunk 1]
```diff
@@ -31,17 +31,16 @@ public IEnumerable<HarvestShapeHit> HarvestShape(HarvestShapeInfo info)
...
-        {
-            var leader = "";
-            if (subPath.StartsWith("Views/") && subPath != "Views/Items")
-            {
-                leader = subPath.Substring("Views/".Length) + "_";
-            }
+        public virtual bool InDepth => false;
 
+        protected virtual string leader(string subPath) => 
+            (subPath.StartsWith("Views/") && subPath != "Views/Items") ? subPath.Substring("Views/".Length) + "_" : "";
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapeTemplateStrategy/BasicShapeTemplateHarvester.cs:37]: `GetLeader()`

`subPath.StartsWith("Views/", StringComparison.Ordinal)` as it is now done in the current dev branch

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapeTemplateStrategy/BasicShapeTemplateHarvester.cs [hunk 2]
```diff
@@ -31,17 +31,16 @@ public IEnumerable<HarvestShapeHit> HarvestShape(HarvestShapeInfo info)
...
+
+        string Adjust(string subPath, string fileName, string displayType)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapeTemplateStrategy/BasicShapeTemplateHarvester.cs:39]: `private string Adjust()`

Currently, in the dev branch, it is `private static string Adjust()` but here can't be static

**Comment by jtkech** [comment]
@aodl @agriffard 

I did a little review

Needs to solve the conflict and take into account the changes that have been done to `BasicShapeTemplateHarvester` in the meantime, e.g. the usage of `StringComparison.Ordinal` and the `private` modifier

Otherwise LGTM

**Comment by agriffard** [comment]
Conflicts need to be solved.

**Comment by agriffard** [comment]
Closing because no answer. Reopen if needed.

**Pull request closed without merging**